experiments/storm/src/main/scala/io/gearpump/experiments/storm/util/StormUtil.scala

```diff
@@ -75,4 +81,185 @@ object StormUtil {
+    grouping.getSetField == Grouping._Fields.FIELDS &&
+        grouping.get_fields.isEmpty
+  }
+
+  def getParallelism(component: ComponentCommon): Int = {
+    val parallelism = component.get_parallelism_hint()
+    if (parallelism == 0) {
+      // for global grouping
+      1
+    } else {
+      parallelism
+    }
+  }
+
+  def getPartitions(msg: Message, target: String): List[Int] = {
+    val stormTuple = msg.msg.asInstanceOf[StormTuple]
+    stormTuple.targetPartitions.getOrElse(target, List(Partitioner.UNKNOWN_PARTITION_ID))
+  }
+
+  def instantiateJavaObject(javaObject: JavaObject): CustomStreamGrouping = {
+    val className = javaObject.get_full_class_name()
+    val args = javaObject.get_args_list().map(_.getFieldValue)
+    val customGrouping = Class.forName(className).getConstructor(args.map(_.getClass): _*)
+        .newInstance(args).asInstanceOf[CustomStreamGrouping]
+    customGrouping
+  }
+
+  def isTickTuple(tuple: Tuple): Boolean = {
+    (tuple.getSourceComponent == Constants.SYSTEM_COMPONENT_ID) && (tuple.getSourceStreamId == Constants.SYSTEM_TICK_STREAM_ID)
+  }
+
+  // a workaround to support storm ShellBolt
+  def mkPidDir: String = {
+    val pidDir = FileUtils.getTempDirectoryPath + File.separator + "pid"
+    try {
+      FileUtils.forceMkdir(new File(pidDir))
+    } catch {
+      case ex: IOException =>
+        LOG.error(s"failed to create pid directory $pidDir")
+    }
+    pidDir
+  }
+
+  // a workaround to support storm ShellBolt
+  def mkCodeDir: String = {
+    val jarPath = System.getProperty("java.class.path").split(":").last
+    val destDir = FileUtils.getTempDirectoryPath + File.separator + "storm"
+
+    try {
+      FileUtils.forceMkdir(new File(destDir))
+
+      val jar = new JarFile(jarPath)
+      val enumEntries = jar.entries()
+      enumEntries.foreach { entry =>
+        val file = new File(destDir + File.separator + entry.getName)
+        if (!entry.isDirectory) {
+          file.getParentFile.mkdirs()
+
+          val is = jar.getInputStream(entry)
+          val fos = new FileOutputStream(file)
+          try {
+            IOUtils.copy(is, fos)
+          } catch {
+            case ex: IOException =>
+              LOG.error(s"failed to copy data from ${entry.getName} to ${file.getName}")
+          } finally {
+            fos.close()
+            is.close()
+          }
+        }
+      }
+    } catch {
+      case ex: IOException =>
+        LOG.error(s"could not extract $destDir from $jarPath")
+    }
+
+    destDir + File.separator + "resources"
+  }
+
+  def buildTopologyContext(topology: StormTopology, stormConf: JMap[_, _],
+      taskToComponent: JMap[Integer, String], componentToSortedTasks: JMap[String, JList[Integer]],
+      componentToStreamFields: JMap[String, JMap[String, Fields]],
+      componentId: String, stormTaskId: Int, multiLang: Boolean = false): TopologyContext = {
+    var codeDir: String = null
+    var pidDir: String = null
+    if (multiLang) {
```

Comment:
multilang needs `codeDir` for source codes for other languages like `splitsentence.py` and `pidDir` for subprocess to run those codes. users shoud not see them. I'd better hide those fields at all.